src/Common/src/TypeSystem/IL/McgInteropSupport.cs

```diff
@@ -0,0 +1,87 @@
+    /// </summary>
+    public static class McgInteropSupport
+    {
+        /// <summary>
+        /// Assembly name suffix for Mcg generated interop code.
+        /// </summary>
+        private const string AssemblyNameSuffix = ".McgInterop";
+
+        /// <summary>
+        /// Returns true if <paramref name="method"/> is Mcg generated PInvoke
+        /// </summary>
+        public static bool IsPregeneratedPInvoke(MethodDesc method)
+        {
+            Debug.Assert(method.IsPInvoke);
+
+            var metadataType = (MetadataType)method.OwningType;
+            var module = metadataType.Module;
+
+            var assemblyName = ((IAssemblyDesc)module).GetName();
+            var simpleName = assemblyName.Name;
+
+            return simpleName.EndsWith(AssemblyNameSuffix);
+        }
+
+        /// <summary>
+        /// Returns pregenerated interop code for given PInvoke method if one exist
+        /// </summary>
+        public static MethodDesc TryGetPregeneratedPInvoke(MethodDesc method)
+        {
+            Debug.Assert(method.IsPInvoke);
+
+            var metadataType = (MetadataType)method.OwningType;
+            var module = metadataType.Module;
+
+            var assemblyName = ((IAssemblyDesc)module).GetName();
+
+            var interopAssemblyName = new AssemblyName();
+
+            interopAssemblyName.Name = assemblyName.Name + AssemblyNameSuffix;
+            interopAssemblyName.Version = assemblyName.Version;
+            interopAssemblyName.SetPublicKeyToken(interopAssemblyName.GetPublicKeyToken());
+            interopAssemblyName.CultureName = assemblyName.CultureName;
+            interopAssemblyName.ContentType = assemblyName.ContentType;
+
+            var interopModule = module.Context.ResolveAssembly(interopAssemblyName, false);
+            if (interopModule == null)
+                return null;
+
+            var generated = GetMatchingMethod(interopModule, method);
+            Debug.Assert(generated != null, "Missing method in .McgInterop assembly" + method.ToString());
+            return generated;
+        }
+
+        private static MethodDesc GetMatchingMethod(ModuleDesc module, MethodDesc method)
+        {
+            var matchingType = GetMatchingType(module, method.OwningType);
+            return matchingType.GetMethod(method.Name, method.Signature);
+        }
+
+        private static TypeDesc GetMatchingType(ModuleDesc module, TypeDesc type)
+        {
+            var metadataType = (MetadataType)type;
+            var containingType = metadataType.ContainingType;
+            if (containingType != null)
+            {
+                var matchingContainingType = (MetadataType)GetMatchingType(module, containingType);
+                return matchingContainingType.GetNestedType(metadataType.Name);
+            }
+            else
+            {
+                return module.GetType(metadataType.Namespace, metadataType.Name);
```

Comment:
There are some inconsistencies in how we handle missing things:
- `module.GetType` will throw if the type is missing
- `matchingContainingType.GetNestedType` will return null and we will nullref later.
- But if the method is missing, we just assert and move on with life.

It would be nice to make it consistent. I would lean towards propagating the null without any throws.